[IMP][12.0] pos_margin: display margin on front end

legalsylvain opened the pull request:
* main feature : Add margin and margin rate on the PoS front end.

![image](https://user-images.githubusercontent.com/3407482/121707543-f3fa2080-cad6-11eb-818b-b22428346584.png)

* Make the summary homogeneous when other OCA modules are installed and fix inheritance (``pos_order_mgmt`` and ``pos_meal_voucher``)

CC : @quentinDupont, @ivantodorovich 

Ref : apps/deck/#/board/144/card/1628

legalsylvain committed:
[IMP] pos_margin : display margin and margin rate in front-end UI

Co-authored-by: Iván Todorovich <ivan.todorovich@gmail.com>

legalsylvain committed:
[REF] pos_meal_voucher : make module compatible with pos_margin. homogenizedisplay of the order summary

legalsylvain committed:
[REF] pos_order_mgmt : make module compatible with pos_margin. homogenizedisplay of the order summary

ivantodorovich commented on pos_margin/static/src/js/models.js:
```diff
@@ -0,0 +1,56 @@
+            var margin = 0;
+            this.orderlines.each(function(orderline){
+                margin += orderline.get_margin();
+            })
```
```suggestion
            return this.get_orderlines().reduce(
                (margin, line) => margin += line.get_margin(), 0
            )
```

This achieves the same
not blocking though

ivantodorovich commented on pos_margin/static/src/js/models.js:
```diff
@@ -0,0 +1,56 @@
+
+        get_margin_rate: function(){
+            var priceWithoutTax = this.get_total_without_tax();
+            if (priceWithoutTax) {
+                return ((this.get_margin()) / priceWithoutTax) * 100;
+            }
+            else {
+                return 0;
+            }
```
```suggestion
            return priceWithoutTax ? ((this.get_margin()) / priceWithoutTax) * 100 : 0;
```
also not blocking

ivantodorovich commented on pos_margin/static/src/js/screens.js:
```diff
@@ -0,0 +1,22 @@
+            this._super.apply(this, arguments);
+            var order = this.pos.get_order();
+            if (!order.get_orderlines().length) {
+                return;
```
I haven't tested, but isn't it necessary to clear the summary in case you remove all order lines?

legalsylvain replied:
no, the summary is hidden if nolines.

length === 0

![image](https://user-images.githubusercontent.com/3407482/121523048-10725c00-c9f6-11eb-9302-55060221005d.png)

length !== 0

![image](https://user-images.githubusercontent.com/3407482/121523126-2849e000-c9f6-11eb-8b76-e39cb1e67eda.png)

ivantodorovich reviewed (approved):
Minor not blocking comments

I would make this new feature configurable through the `pos.config`, and disabled by default.

I'm thinking current users of this module might not want to display the margin to their cashiers, where customers can potentially see it.

legalsylvain commented:
1) Thanks for your review !

2) thanks for the JS suggestion, I included it in the PR.
 
3)
> I would make this new feature configurable through the pos.config, and disabled by default. I'm thinking current users of this module might not want to display the margin to their cashiers, where customers can potentially see it.

OK to make it configurable, you're right depending on the configuration of the shop, the values could be visible by the customer, and in some cases, this could be embarrassing.

- But, I'd like to set it enabled by default because in this way, all the features are enabled and admin can disable some ones, if required. Otherwise admin could not know that the feature is available and could miss this feature.
- I think that not only code should be open, margin should also be Open Source. In grap, we practice price transparency. (but that is political choice) I know that in food coops, that is the same. (Park Slope Food Coop at New York, La Louve at Paris, etc..)

If it's a problem for existing instance, I can simply add a migration script to disable the feature.

could be OK ? 

Note : Context of the development : users don't allways set correctly the standard_price on each product form. So the margin is bad computed. It is more a new feature to check that your products are correctly set and fix quickly bad values.

ivantodorovich commented:
Personally I'm not using this module, so I'm not blocking over this, but I'm thinking on the people that have this module installed, decided to update it and unknowingly started to display cost & margin in the Point of Sale.

A migration script to disable the feature in existing installations would cover this case, if you want it enabled by default.

legalsylvain commented:
Hi @ivantodorovich 

- I added a configuration option in the ``pos.config``

![pos_config_form](https://user-images.githubusercontent.com/3407482/121708345-c19cf300-cad7-11eb-84d3-b94a52eb7437.png)


- I added a migration script to disable that feature, for existing instance
- I udpated the description and add ``CONFIGURE.rst`` section
- I added the margin rate on each line below the price (see updated print screen)


**merge should be done with nobump option**

legalsylvain commented:
/ocabot merge nobump
